File: rubygems/pray-cli/lib/pray/torrent_manifest.rb
# frozen_string_literal: true

require "json"
require "fileutils"

module Pray
  TORRENT_MANIFEST_SPEC = "pray-torrent-v1"
  DEFAULT_TORRENT_PIECE_SIZE = 16 * 1024

  module TorrentManifest
    module_function

    def path_for(artifact_path)
      "#{artifact_path}.praytorrent.json"
    end

    def payload(name:, version:, artifact_path:, archive_bytes:, trackers: [])
      {
        "spec" => TORRENT_MANIFEST_SPEC,
        "name" => name,
        "version" => version,
        "artifact_url" => artifact_path,
        "artifact_hash" => Hashing.sha256_prefixed(archive_bytes),
        "piece_size" => DEFAULT_TORRENT_PIECE_SIZE,
        "length" => archive_bytes.bytesize,
        "pieces" => piece_hashes(archive_bytes),
        "sources" => [artifact_path],
        "trackers" => Array(trackers)
      }
    end

    def bytes(name:, version:, artifact_path:, archive_bytes:, trackers: [])
      JSON.pretty_generate(
        payload(
          name: name,
          version: version,
          artifact_path: artifact_path,
          archive_bytes: archive_bytes,
          trackers: trackers
        )
      )
    end

    def write(root, name:, version:, artifact_path:, archive_bytes:, settings:)
      return unless settings.allows_torrent?

      path = File.join(root, path_for(artifact_path))
      FileUtils.mkdir_p(File.dirname(path))
      File.write(
        path,
        bytes(
          name: name,
          version: version,
          artifact_path: artifact_path,
          archive_bytes: archive_bytes,
          trackers: settings.bootstrap_trackers
        )
      )
    end

    def piece_hashes(archive_bytes)
      length = archive_bytes.bytesize
      return [] if length.zero?

      hashes = []
      start = 0
      while start < length
        hashes << Hashing.sha256_prefixed(archive_bytes.byteslice(start, DEFAULT_TORRENT_PIECE_SIZE))
        start += DEFAULT_TORRENT_PIECE_SIZE
      end
      hashes
    end

    def descriptor_present?(root, artifact_path, settings)
      !settings.allows_torrent? || File.file?(File.join(root, path_for(artifact_path)))
    end
  end
end

File: npmjs/pray-cli/src/publish/torrent-manifest.ts
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import {
  allowsTorrent,
  defaultDistributionSettings,
  parseDistributionSettings,
  type RegistryDistributionSettings,
} from "../distribution.js";
import { PrayError } from "../errors.js";
import { sha256Prefixed } from "../hashing.js";
import { httpGetText, joinUrl } from "../http/client.js";

export const TORRENT_MANIFEST_SPEC = "pray-torrent-v1";
export const DEFAULT_TORRENT_PIECE_SIZE = 16 * 1024;

export interface TorrentManifestPayload {
  spec: string;
  name: string;
  version: string;
  artifact_url: string;
  artifact_hash: string;
  piece_size: number;
  length: number;
  pieces: string[];
  sources: string[];
  trackers: string[];
}

export function torrentManifestPath(artifactPath: string): string {
  return `${artifactPath}.praytorrent.json`;
}

export function torrentManifestPayload(
  name: string,
  version: string,
  artifactPath: string,
  archiveBytes: Buffer,
  trackers: string[],
): TorrentManifestPayload {
  return {
    spec: TORRENT_MANIFEST_SPEC,
    name,
    version,
    artifact_url: artifactPath,
    artifact_hash: sha256Prefixed(archiveBytes),
    piece_size: DEFAULT_TORRENT_PIECE_SIZE,
    length: archiveBytes.length,
    pieces: pieceHashes(archiveBytes),
    sources: [artifactPath],
    trackers,
  };
}

export function torrentManifestBytes(
  name: string,
  version: string,
  artifactPath: string,
  archiveBytes: Buffer,
  trackers: string[],
): string {
  return `${JSON.stringify(
    torrentManifestPayload(name, version, artifactPath, archiveBytes, trackers),
    null,
    2,
  )}\n`;
}

export function writeTorrentManifest(
  root: string,
  name: string,
  version: string,
  artifactPath: string,
  archiveBytes: Buffer,
  settings: RegistryDistributionSettings,
): void {
  if (!allowsTorrent(settings)) return;
  const path = join(root, torrentManifestPath(artifactPath));
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(
    path,
    torrentManifestBytes(
      name,
      version,
      artifactPath,
      archiveBytes,
      settings.bootstrapTrackers,
    ),
  );
}

export async function fetchDistributionSettings(
  sourceUrl: string,
): Promise<RegistryDistributionSettings> {
  try {
    return parseDistributionSettings(
      await httpGetText(joinUrl(sourceUrl, "v1/distribution.json")),
    );
  } catch (error) {
    if (isHttpNotFound(error)) {
      return defaultDistributionSettings();
    }
    throw error;
  }
}

export function torrentDescriptorPresent(
  root: string,
  artifactPath: string,
  settings: RegistryDistributionSettings,
): boolean {
  return (
    !allowsTorrent(settings) ||
    existsSync(join(root, torrentManifestPath(artifactPath)))
  );
}

function pieceHashes(archiveBytes: Buffer): string[] {
  if (archiveBytes.length === 0) return [];
  const hashes: string[] = [];
  for (
    let start = 0;
    start < archiveBytes.length;
    start += DEFAULT_TORRENT_PIECE_SIZE
  ) {
    hashes.push(
      sha256Prefixed(
        archiveBytes.subarray(start, start + DEFAULT_TORRENT_PIECE_SIZE),
      ),
    );
  }
  return hashes;
}

function isHttpNotFound(error: unknown): boolean {
  return error instanceof PrayError && /:\s*404\b/.test(error.message);
}
